data/src/message.rs

```diff
@@ -173,19 +191,168 @@ pub enum Direction {
     Received,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Reaction {
+    pub target: Target,
+    pub sender: Nick,
+    pub text: String,
+    pub id: Option<Id>,
+    pub in_reply_to: Id,
+    pub unreact: bool,
+}
+
+impl Reaction {
+    pub fn with_target(self, target: Target) -> Self {
+        Self { target, ..self }
+    }
+    pub fn sent(
+        target: Target,
+        my_nick: Nick,
+        in_reply_to: Id,
+        text: String,
+        unreact: bool,
+    ) -> Self {
+        Self {
+            sender: my_nick,
+            target,
+            in_reply_to,
+            text,
+            id: None,
+            unreact,
+        }
+    }
+    pub fn unreact(
+        target: Target,
+        my_nick: Nick,
+        in_reply_to: Id,
+        text: String,
+    ) -> Self {
+        Self {
+            sender: my_nick,
+            target,
+            in_reply_to,
+            text,
+            id: None,
+            unreact: true,
+        }
+    }
+}
+
 #[derive(Debug, Clone)]
 pub struct Message {
     pub received_at: Posix,
     pub server_time: DateTime<Utc>,
     pub direction: Direction,
     pub target: Target,
     pub content: Content,
-    pub id: Option<String>,
+    pub id: Option<Id>,
     pub hash: Hash,
     pub hidden_urls: HashSet<Url>,
     pub is_echo: bool,
 }
 
+pub enum Decoded {
+    Message(Message),
+    Reaction(Reaction),
+}
+impl Decoded {
+    pub fn into_highlight(
+        &self,
+        server: &Server,
+    ) -> Option<(Message, Channel, User)> {
+        match self {
+            Self::Message(message) => message.into_highlight(server),
+            _ => None,
+        }
+    }
+    pub fn with_target(self, target: Target) -> Self {
+        match self {
+            Self::Message(message) => {
+                Self::Message(message.with_target(target))
+            }
+            Self::Reaction(reaction) => {
+                Self::Reaction(reaction.with_target(target))
+            }
+        }
+    }
+    pub fn target(&self) -> &Target {
+        match self {
+            Self::Message(message) => &message.target,
+            Self::Reaction(reaction) => &reaction.target,
+        }
+    }
```

Comment:
This method isn't needed. It's only called on a code path where we de-structure to `Message`, which it can just be referenced from that.